src/devices/cpu/cadr/cadr_dasm.cpp

```diff
@@ -0,0 +1,600 @@
+// license:BSD-3-Clause
+// copyright-holders:Wilbert Pol
+/***************************************************************************
+
+    MIT CADR microcode disassembler
+
+The official assembler language is unknown.
+An attempt has been made to keep the generated assembler code short and
+understandable.
+
+TODO:
+- bit 45 ilong
+
+***************************************************************************/
+
+#include "emu.h"
+#include "cadr_dasm.h"
+
+
+const char *const cadr_disassembler::bool_op[0x10] =
+{
+	"SETZ",  "AND", "ANDCA", "SETM", "ANDCM", "SETA", "XOR",  "IOR",
+	"ANDCB", "EQV", "SETCA", "ORCA", "SETCM", "ORCM", "ORCB", "SETO"
+};
+
+const char *const cadr_disassembler::arith_op[0x10] =
+{
+	"-1",  "(M&A)-1", "(M&~A)-1",     "M-1",     "M|~A", "(M|~A)+(M&A)", "M-A-1",    "(M|~A)+M",
+	"M|A", "M+A",     "(M|A)+(M&~A)", "(M|A)+M", "M",    "M+(M&A)",      "M+(M|~A)", "M+M"
+};
+
+const char *const cadr_disassembler::arith_op_c[0x10] =
+{
+	"0",       "M&A",   "(M&~A)",         "M",         "(M|~A)+1", "(M|~A)+(M&A)+1", "M-A",        "(M|~A)+M+1",
+	"(M|A)+1", "M+A+1", "(M|A)+(M&~A)+1", "(M|A)+M+1", "M+1",      "M+(M&A)+1",      "M+(M|~A)+1", "M+M+1"
+};
+
+const char *const cadr_disassembler::mult_div_op[0x20] =
+{
+	"mult-step", "div-step",      "unk", "unk", "unk", "rem-corr", "unk", "unk",
+	"unk",       "init-div-step", "unk", "unk", "unk", "unk",      "unk", "unk",
+	"mult-step", "div-step",      "unk", "unk", "unk", "rem-corr", "unk", "unk",
+	"unk",       "init-div-step", "unk", "unk", "unk", "unk",      "unk", "unk"
+};
+
+const char *const cadr_disassembler::output_bus_control[0x04] =
+{
+	"ill ", "", ">> ", "<< "
+};
+
+const char *const cadr_disassembler::q_control[0x04] =
+{
+	"", "<<Q,", ">>Q,", "Q,"
+};
+
+const char *const cadr_disassembler::rp[0x04] =
+{
+	"branch", "call", "return", "imem-write"
+};
+
+
+u32 cadr_disassembler::opcode_alignment() const
+{
+	return 1;
+}
+
+void cadr_disassembler::a_source(std::ostream &stream, u64 op)
+{
+	util::stream_format(stream, "a[%o]", ((op >> 32) & 0x3ff));
+}
+
+
+void cadr_disassembler::m_source(std::ostream &stream, u64 op)
```

Comment:
These helper functions that don’t depend on the disassembler object state at all can be turned into free functions in an anonymous namespace to keep implementation details out of the header.